rfcs/shared/build-system/package-public-api.md

```diff
@@ -0,0 +1,237 @@
+
+## Problem statement
+
+We expose all APIs of our packages as a PUBLIC from barrel files (`src/index.ts`) to consumers, which introduces problems as we can expose apis that were supposed to be used only internally within our repo or only be available in particular pre-release stage (ALPHA/BETA versions).
+
+## Detailed Design or Proposal
+
+We wanna provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.
+
+Note that our approach will be applied only on **Type Declaration emit level** as we consider TypeScript surface as only source of truth for our package APIs.
+
+> This means that if users use only vanilla JavaScript nothing will stop them from using those API's - as we don't support that kind of usage users take the risk that things may/will break for them unfortunately and also they will not get any support from our side.
+
+### Proposed access modifiers:
+
+#### `@internal`
+
+It indicates that an API item is meant only for usage by other NPM packages from the same repo. Third parties _should never use “internal” APIs._
+
+- tooling will trim the declaration from a public API surface.
+- tooling will trim the declaration from `/unstable` API surface.
+
+**Example:**
+
+```ts
+// @filename package-a/src/index.ts
+
+/**
+ *
+ * @internal
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+
+export declare function hello(): void;
+```
+
+#### `@alpha`
+
+It indicates that an API item is eventually intended to be public, but currently is in an early stage of development
+
+- tooling will trim the declaration from a public API surface.
+- can be exposed only from `/unstable` API surface
+- tolling will NOT TRIM the declaration from `/unstable` API surface.
+
+**Example:**
+
+- unstable api
+
+```ts
+// @filename package-a/src/unstable/index.ts
+
+/**
+ *
+ * @alpha
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/unstable.d.ts
+export function doSomething(): void;
+export declare function hello(): void;
+```
+
+- stable api
+
+```ts
+// @filename package-a/src/index.ts
+
+// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
+/**
+ *
+ * @alpha
+ */
+export function doSomething(): void {}
+/**
+ *
+ * @public
+ */
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+export declare function hello(): void;
+```
+
+#### `@beta`
```

Comment:
I don't see any difference between in `@alpha` & `@beta` except semantics, is correct? As we export them as under unstable is there a sense to use `@unstable` instead?

IMO it will reduce maintenance cost as with current proposal we should ensure that `@alpha` will be changed to `@beta` when a package will be promoted. And we don't have any automation for it...

Reply:
> I don't see any difference between in @alpha & @beta except semantics, is correct 

correct. I added more context to one of Lings comments
↓
![image](https://user-images.githubusercontent.com/1223799/174262747-0a5d0cd0-e595-4346-bcf2-209e58b0e4ce.png)

> As we export them as under unstable is there a sense to use @unstable instead?

that's great idea ! we would need to get this implemented to api-extractor / or eventually introduce custom tooling

Reply:
https://github.com/microsoft/rushstack/issues/3558